rasa/core/training/structures.py

```diff
@@ -400,6 +400,22 @@ def __init__(
         self.story_steps = story_steps
         self.step_lookup = {s.id: s for s in self.story_steps}
         ordered_ids, cyclic_edges = StoryGraph.order_steps(story_steps)
+
+        # sort ordered_ids for reproducible results
+        ordered_ids = list(ordered_ids)
+        ordered_ids.sort(key=lambda x: int(x.split("_")[0]))
+        ordered_ids = deque(ordered_ids)
+
+        # sort cyclic edges for reproducible results
+        cyclic_edges = list(cyclic_edges)
+        cyclic_edges = [
+            (x, y) if int(x.split("_")[0]) < int(y.split("_")[0]) else (y, x)
+            for x, y in cyclic_edges
+        ]
```

Comment:
As the `cyclic_edges` are directed, they should not be changed.

Reply:
@tabergma In theory, I agree with you. But when i print cyclic_edges for 2 consecutive runs i get reversed edges. Following are the outputs:
cyclic_edges from run 1:
{('15_9fb1c8d31bc844a591f179e02333a848','15_9fb1c8d31bc844a591f179e02333a848'), ('11_60a242338be345c6802ccb23cd6eb133', '15_9fb1c8d31bc844a591f179e02333a848'), ('3_2e32f36501cb47438975fe4f3dd0509b', '15_9fb1c8d31bc844a591f179e02333a848'), ('7_adfdb00270f948fe8e2be31747308fb8', '15_9fb1c8d31bc844a591f179e02333a848'), ('7_adfdb00270f948fe8e2be31747308fb8', '11_60a242338be345c6802ccb23cd6eb133'), ('7_adfdb00270f948fe8e2be31747308fb8', '3_2e32f36501cb47438975fe4f3dd0509b'), ('3_2e32f36501cb47438975fe4f3dd0509b', '11_60a242338be345c6802ccb23cd6eb133'), ('11_60a242338be345c6802ccb23cd6eb133', '11_60a242338be345c6802ccb23cd6eb133'), ('3_2e32f36501cb47438975fe4f3dd0509b', '3_2e32f36501cb47438975fe4f3dd0509b'), ('21_7e1e88dc81ef47ecb6397d71f0772579', '21_7e1e88dc81ef47ecb6397d71f0772579'), ('7_adfdb00270f948fe8e2be31747308fb8', '21_7e1e88dc81ef47ecb6397d71f0772579'), ('21_7e1e88dc81ef47ecb6397d71f0772579', '11_60a242338be345c6802ccb23cd6eb133'), ('21_7e1e88dc81ef47ecb6397d71f0772579', '15_9fb1c8d31bc844a591f179e02333a848'), ('21_7e1e88dc81ef47ecb6397d71f0772579', '3_2e32f36501cb47438975fe4f3dd0509b'), ('7_adfdb00270f948fe8e2be31747308fb8', '7_adfdb00270f948fe8e2be31747308fb8')}
cyclic_edges from run 2:
{('7_eea69044a80c44c897ba4f79475267ed', '7_eea69044a80c44c897ba4f79475267ed'), ('3_c0afc5a2289042d3912446844bb3ea34', '7_eea69044a80c44c897ba4f79475267ed'), ('15_133d69532c42460db4326baf68ad7303', '15_133d69532c42460db4326baf68ad7303'), ('21_c8ba96bc4e7f4fe2b823b0779326684b', '7_eea69044a80c44c897ba4f79475267ed'), ('15_133d69532c42460db4326baf68ad7303', '3_c0afc5a2289042d3912446844bb3ea34'), ('15_133d69532c42460db4326baf68ad7303', '7_eea69044a80c44c897ba4f79475267ed'), ('3_c0afc5a2289042d3912446844bb3ea34', '3_c0afc5a2289042d3912446844bb3ea34'), ('11_49f9703444bf4b979839a718a0b4e42d', '3_c0afc5a2289042d3912446844bb3ea34'), ('11_49f9703444bf4b979839a718a0b4e42d', '11_49f9703444bf4b979839a718a0b4e42d'), ('21_c8ba96bc4e7f4fe2b823b0779326684b', '3_c0afc5a2289042d3912446844bb3ea34'), ('21_c8ba96bc4e7f4fe2b823b0779326684b', '11_49f9703444bf4b979839a718a0b4e42d'), ('21_c8ba96bc4e7f4fe2b823b0779326684b', '21_c8ba96bc4e7f4fe2b823b0779326684b'), ('15_133d69532c42460db4326baf68ad7303', '11_49f9703444bf4b979839a718a0b4e42d'), ('15_133d69532c42460db4326baf68ad7303', '21_c8ba96bc4e7f4fe2b823b0779326684b'), ('11_49f9703444bf4b979839a718a0b4e42d', '7_eea69044a80c44c897ba4f79475267ed')}

We can notice that in first output we have (s,e)=(11_,15_) while in second output we have (s,e)=(15_,11_) whereas (21_,3_) and (21_,11_) edges remain unchanged. Please let me know if I am misinterpreting. I believe this behaviour can be reproduced on any dataset with cycles.

Reply:
the first number in the id is step count, which doesn't always correspond to topological order. That's why we have an algorithm [`topological_sort`](https://github.com/RasaHQ/rasa/blob/8775862ea6444ef47402c67c40aacb8162efdb36/rasa/core/training/structures.py#L656) that tries to order directed graph.

Reply:
is it not possible to stabilise topological_sort by changing the graph data structure from graph: Dict[Text, Set[Text]] to graph: Dict[Text, list[Text]]?

Reply:
Why would be stabilized? the `Set` there means that a node cannot be connected to another node with two edges

Reply:
The main source of instability of topological_sort is the set() ds which is unordered. 
In rasa/core/training/structures.py; function dfs(); line 688 we have "for k in graph.get(node, set()):" because graph.get(node, set()) is a set() it is ordered differently in every run thus, causing random graph traversals in dfs(). If instead we use "for k in sorted(graph.get(node, set())):" the order of traversal becomes fixed but the algorithm still remains intact. This change will get us deterministic traversal and deterministic "ordered" variable. Thus, no need to sort ordered_ids afterwards.

Due to the above change, the order of (s,e) pairs in removed_edges variables also becomes fixed and does not change across runs as reported above in this thread. But for removed_edges we are again using set() ds making it unordered. Thus, we need to make a "removed_edges = sorted(removed_edges)" call to fix the order. In my opinion, these changes are not breaking the algorithm and stabilising topological_sort function. Making these changes for review

Reply:
hm, the problem sounds reasonable, the only problem I see in this case is that then we prefer one topology against the other, without any good reason. I'll give it a thought, thank you

Reply:
@tmbo what do you think?